File: ruby/beer-song/beer.rb
class Beer
  def verse(number)
    Verse.new(number).to_s
  end

  def sing(beginning, ending = 0)
    verses(beginning, ending).join("\n") + "\n"
  end

  private

  def verses(beginning, ending)
    beginning.downto(ending).map do |number|
      Verse.new(number).to_s
    end
  end

  class Verse
    def initialize(number)
      @remaining = number
    end

    def to_s
      "#{first_line}\n#{second_line}\n"
    end

    private

    attr_reader :remaining

    def first_line
      "#{bottles(remaining).capitalize} of beer on the wall, #{bottles(remaining)} of beer."
    end

    def second_line
      line =
        if out?
          'Go to the store and buy some more'
        else
          "Take #{last_one? ? 'it' : 'one'} down and pass it around"
        end

      line << ", #{bottles(remaining - 1)} of beer on the wall."
    end

    def out?
      remaining.zero?
    end

    def last_one?
      remaining == 1
    end

    def bottles(remaining)
      case remaining
      when -1
        '99 bottles'
      when 0
        'no more bottles'
      when 1
        '1 bottle'
      else
        "#{remaining} bottles"
      end
    end
  end
end

File: elixir/beer-song/beer.exs
defmodule Beer do
  def verse(bottles) do
    first_line(bottles) <> second_line(bottles)
  end

  def sing(beginning, ending // 0) when beginning == ending do
    verse(beginning) <> "\n"
  end
  def sing(beginning, ending) do
    Enum.join([verse(beginning), sing(beginning - 1, ending)], "\n")
  end

  defp first_line(remaining) do
    "#{String.capitalize(bottles(remaining))} of beer on the wall, #{bottles(remaining)} of beer.\n"
  end

  defp second_line(remaining) do
    line = if remaining == 0 do
      "Go to the store and buy some more"
    else
      noun = if remaining == 1, do: 'it', else: 'one'

      "Take #{noun} down and pass it around"
    end

    "#{line}, #{bottles(if remaining == 0, do: 99, else: remaining - 1)} of beer on the wall.\n"
  end

  defp bottles(0),         do: "no more bottles"
  defp bottles(1),         do: "1 bottle"
  defp bottles(remaining), do: Enum.join([remaining, "bottles"], " ")
end
